File: Drop-In/src/test/java/com/braintreepayments/api/DropInLifecycleObserverUnitTest.java
package com.braintreepayments.api;

import static org.mockito.Matchers.any;
import static org.mockito.Matchers.anyString;
import static org.mockito.Matchers.eq;
import static org.mockito.Matchers.same;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.verify;

import androidx.activity.result.ActivityResultCallback;
import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.ActivityResultRegistry;
import androidx.fragment.app.FragmentActivity;
import androidx.lifecycle.Lifecycle;
import androidx.lifecycle.LifecycleOwner;

import junit.framework.TestCase;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.ArgumentCaptor;
import org.mockito.Captor;
import org.mockito.Mockito;
import org.mockito.MockitoAnnotations;
import org.robolectric.RobolectricTestRunner;

@RunWith(RobolectricTestRunner.class)
public class DropInLifecycleObserverUnitTest extends TestCase {

    @Captor
    ArgumentCaptor<ActivityResultCallback<DropInResult>> dropInResultCaptor;

    @Before
    public void beforeEach() {
        MockitoAnnotations.initMocks(this);
    }

    @Test
    public void onCreate_registersForAnActivityResult() {
        ActivityResultRegistry activityResultRegistry = mock(ActivityResultRegistry.class);
        DropInClient dropInClient = mock(DropInClient.class);
        DropInLifecycleObserver sut =
                new DropInLifecycleObserver(activityResultRegistry, dropInClient);

        FragmentActivity lifecycleOwner = new FragmentActivity();
        sut.onStateChanged(lifecycleOwner, Lifecycle.Event.ON_CREATE);

        String expectedKey = "com.braintreepayments.api.DropIn.RESULT";
        verify(activityResultRegistry).register(eq(expectedKey), same(lifecycleOwner), any(DropInActivityResultContract.class), Mockito.any());
    }

    @Test
    public void onCreate_whenActivityResultReceived_forwardsResultToDropInClient() {
        ActivityResultRegistry activityResultRegistry = mock(ActivityResultRegistry.class);
        DropInClient dropInClient = mock(DropInClient.class);
        DropInLifecycleObserver sut =
                new DropInLifecycleObserver(activityResultRegistry, dropInClient);

        FragmentActivity lifecycleOwner = new FragmentActivity();
        sut.onStateChanged(lifecycleOwner, Lifecycle.Event.ON_CREATE);

        verify(activityResultRegistry).register(anyString(), any(LifecycleOwner.class), any(DropInActivityResultContract.class), dropInResultCaptor.capture());
        ActivityResultCallback<DropInResult> activityResultCallback = dropInResultCaptor.capture();

        DropInResult dropInResult = new DropInResult();
        activityResultCallback.onActivityResult(dropInResult);
        verify(dropInClient).onDropInResult(dropInResult);
    }

    @Test
    public void onCreate_whenActivityResultIsNull_doesNotCallListener() {
        ActivityResultRegistry activityResultRegistry = mock(ActivityResultRegistry.class);
        DropInClient dropInClient = mock(DropInClient.class);
        DropInLifecycleObserver sut =
                new DropInLifecycleObserver(activityResultRegistry, dropInClient);

        FragmentActivity lifecycleOwner = new FragmentActivity();
        sut.onStateChanged(lifecycleOwner, Lifecycle.Event.ON_CREATE);

        verify(activityResultRegistry).register(anyString(), any(LifecycleOwner.class), any(DropInActivityResultContract.class), dropInResultCaptor.capture());
        ActivityResultCallback<DropInResult> activityResultCallback = dropInResultCaptor.capture();

        activityResultCallback.onActivityResult(null);
        verify(dropInClient, never()).onDropInResult(any(DropInResult.class));
    }

    @Test
    public void launch_launchesActivity() {
        DropInRequest dropInRequest = new DropInRequest();
        Authorization authorization = Authorization.fromString(Fixtures.BASE64_CLIENT_TOKEN);

        ActivityResultRegistry activityResultRegistry = mock(ActivityResultRegistry.class);

        DropInClient dropInClient = mock(DropInClient.class);
        DropInLifecycleObserver sut =
                new DropInLifecycleObserver(activityResultRegistry, dropInClient);

        ActivityResultLauncher<DropInIntentData> activityLauncher = mock(ActivityResultLauncher.class);
        sut.activityLauncher = activityLauncher;

        DropInIntentData dropInIntentData =
                new DropInIntentData(dropInRequest, authorization, "sample-session-id");
        sut.launch(dropInIntentData);

        verify(activityLauncher).launch(dropInIntentData);
    }
}

File: Drop-In/src/test/java/com/braintreepayments/api/DropInLifecycleObserverUnitTest.kt
package com.braintreepayments.api

import androidx.activity.result.ActivityResultCallback
import androidx.activity.result.ActivityResultLauncher
import androidx.activity.result.ActivityResultRegistry
import androidx.fragment.app.FragmentActivity
import io.mockk.every
import io.mockk.mockk
import io.mockk.slot
import io.mockk.verify
import junit.framework.TestCase
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner

@RunWith(RobolectricTestRunner::class)
class DropInLifecycleObserverUnitTest : TestCase() {

    @Test
    fun onCreate_registersForAnActivityResult() {
        val activityResultRegistry = mockk<ActivityResultRegistry>(relaxed = true)
        val dropInClient = mockk<DropInClient>(relaxed = true)

        val sut = DropInLifecycleObserver(activityResultRegistry, dropInClient)

        val lifecycleOwner = FragmentActivity()
        sut.onCreate(lifecycleOwner)

        val expectedKey = "com.braintreepayments.api.DropIn.RESULT"
        verify {
            activityResultRegistry.register(
                expectedKey,
                lifecycleOwner,
                any<DropInActivityResultContract>(),
                any()
            )
        }
    }

    @Test
    fun onCreate_whenActivityResultReceived_forwardsResultToDropInClient() {
        val activityResultRegistry = mockk<ActivityResultRegistry>(relaxed = true)
        val dropInClient = mockk<DropInClient>(relaxed = true)

        val callbackSlot = slot<ActivityResultCallback<DropInResult>>()
        val activityLauncher: ActivityResultLauncher<DropInIntentData> = mockk(relaxed = true)
        every {
            activityResultRegistry.register(
                any(),
                any(),
                any<DropInActivityResultContract>(),
                capture(callbackSlot)
            )
        } returns activityLauncher

        val lifecycleOwner = FragmentActivity()
        val sut = DropInLifecycleObserver(activityResultRegistry, dropInClient)
        sut.onCreate(lifecycleOwner)

        val dropInResult = DropInResult()
        callbackSlot.captured.onActivityResult(dropInResult)
        verify { dropInClient.onDropInResult(dropInResult) }
    }

    @Test
    fun launch_launchesActivity() {
        val dropInRequest = DropInRequest()
        val authorization = Authorization.fromString(Fixtures.BASE64_CLIENT_TOKEN)

        val activityResultRegistry = mockk<ActivityResultRegistry>(relaxed = true)
        val dropInClient = mockk<DropInClient>(relaxed = true)

        val activityLauncher: ActivityResultLauncher<DropInIntentData> = mockk(relaxed = true)
        every {
            activityResultRegistry.register(
                any(),
                any(),
                any<DropInActivityResultContract>(),
                any()
            )
        } returns activityLauncher

        val sut = DropInLifecycleObserver(activityResultRegistry, dropInClient)

        val lifecycleOwner = FragmentActivity()
        sut.onCreate(lifecycleOwner)

        val dropInIntentData = DropInIntentData(dropInRequest, authorization, "sample-session-id")
        sut.launch(dropInIntentData)

        verify { activityLauncher.launch(dropInIntentData) }
    }
}